File: example/android/app/src/test/kotlin/com/exponea/example/ConfigurationChangeTest.kt
package com.exponea.example

import com.exponea.data.ExponeaConfigurationParser
import com.exponea.sdk.models.EventType
import com.exponea.sdk.models.ExponeaConfiguration
import org.junit.Assert.assertEquals
import org.junit.Assert.fail
import org.junit.BeforeClass
import org.junit.Test

class ConfigurationChangeTest {
    companion object {
        private val DEFAULT_URL = "default-base-url"

        lateinit var data: List<Map<String, Any?>>

        @BeforeClass @JvmStatic fun setup() {
            data = BaseTest.readMapData("configuration_change")
        }
    }

    @Test
    fun `validate data`() {
        assertEquals(data.size, 4)
    }

    @Test
    fun `parse empty map`() {
        val parser = ExponeaConfigurationParser()
        val decoded = parser.parseConfigChange(data[0], DEFAULT_URL)

        assertEquals(decoded.project, null)
        assertEquals(decoded.mapping, null)
    }

    @Test
    fun `parse minimal`() {
        val parser = ExponeaConfigurationParser()
        val decoded = parser.parseConfigChange(data[1], DEFAULT_URL)

        assertEquals(decoded.project != null, true)
        val project = decoded.project!!
        assertEquals(project.projectToken, "mock-project-token")
        assertEquals(project.authorization, "Token mock-auth-token")
        assertEquals(project.baseUrl, DEFAULT_URL)
        assertEquals(decoded.mapping, null)
    }

    @Test
    fun `parse with base url`() {
        val parser = ExponeaConfigurationParser()
        val decoded = parser.parseConfigChange(data[2], DEFAULT_URL)

        assertEquals(decoded.project != null, true)
        val project = decoded.project!!
        assertEquals(project.projectToken, "mock-project-token")
        assertEquals(project.authorization, "Token mock-auth-token")
        assertEquals(project.baseUrl, "http://mock.base.url.com")
        assertEquals(decoded.mapping, null)
    }

    @Test
    fun `parse full`() {
        val parser = ExponeaConfigurationParser()
        val decoded = parser.parseConfigChange(data[3], DEFAULT_URL)

        assertEquals(decoded.project != null, true)
        val project = decoded.project!!
        assertEquals(project.projectToken, "mock-project-token")
        assertEquals(project.authorization, "Token mock-auth-token")
        assertEquals(project.baseUrl, "http://mock.base.url.com")
        assertEquals(decoded.mapping != null, true)
        val mapping = decoded.mapping!!
        assertEquals(mapping.size, 1);
        val projectList = mapping[EventType.PAYMENT]!!;
        assertEquals(projectList.size, 1);
        assertEquals(projectList[0].projectToken, "other-project-token");
        assertEquals(projectList[0].authorization, "Token other-auth-token");
        assertEquals(projectList[0].baseUrl, project.baseUrl);
    }
}

File: test/unit/data/configuration_change_test.dart
import 'package:exponea/exponea.dart';
import 'package:exponea/src/data/encoder/main.dart';
import 'package:flutter_test/flutter_test.dart';

import 'base.dart';

void main() {
  group('ExponeaConfigurationChange', () {
    const encode = ExponeaConfigurationChangeEncoder.encode;
    const decode = ExponeaConfigurationChangeEncoder.decode;

    final data = readMapData('configuration_change');
    test('check data', () async {
      expect(data.length, 4);
    });
    final noData = data[0];
    final minData = data[1];
    final withBaseUrlData = data[2];
    final fullData = data[3];

    group('encode', () {
      test('minimal', () async {
        const change = ExponeaConfigurationChange(
          project: ExponeaProject(
            projectToken: 'mock-project-token',
            authorizationToken: 'mock-auth-token',
          ),
        );

        expect(encode(change), minData);
      });

      test('with base url', () async {
        const change = ExponeaConfigurationChange(
          project: ExponeaProject(
            projectToken: 'mock-project-token',
            authorizationToken: 'mock-auth-token',
            baseUrl: 'http://mock.base.url.com',
          ),
        );

        expect(encode(change), withBaseUrlData);
      });

      test('full', () async {
        const change = ExponeaConfigurationChange(
          project: ExponeaProject(
            projectToken: 'mock-project-token',
            authorizationToken: 'mock-auth-token',
            baseUrl: 'http://mock.base.url.com',
          ),
          mapping: {
            EventType.payment: [
              ExponeaProject(
                projectToken: 'other-project-token',
                authorizationToken: 'other-auth-token',
              ),
            ],
          },
        );

        expect(encode(change), fullData);
      });
    });

    group('decode', () {
      test('no data', () async {
        final decoded = decode(noData);

        expect(decoded.project != null, false);
        expect(decoded.mapping != null, false);
      });

      test('minimal', () async {
        final decoded = decode(minData);

        expect(decoded.project != null, true);
        expect(decoded.project!.projectToken, 'mock-project-token');
        expect(decoded.project!.authorizationToken, 'mock-auth-token');
        expect(decoded.project!.baseUrl, null);
        expect(decoded.mapping != null, false);
      });

      test('with base url', () async {
        final decoded = decode(withBaseUrlData);

        expect(decoded.project != null, true);
        expect(decoded.project!.projectToken, 'mock-project-token');
        expect(decoded.project!.authorizationToken, 'mock-auth-token');
        expect(decoded.project!.baseUrl, 'http://mock.base.url.com');
        expect(decoded.mapping != null, false);
      });

      test('full', () async {
        final decoded = decode(fullData);

        expect(decoded.project != null, true);
        expect(decoded.project!.projectToken, 'mock-project-token');
        expect(decoded.project!.authorizationToken, 'mock-auth-token');
        expect(decoded.project!.baseUrl, 'http://mock.base.url.com');
        expect(decoded.mapping != null, true);
        final mapping = decoded.mapping!;
        expect(mapping.length, 1);
        final projectList = mapping[EventType.payment]!;
        expect(projectList.length, 1);
        expect(projectList.single.projectToken, 'other-project-token');
        expect(projectList.single.authorizationToken, 'other-auth-token');
        expect(projectList.single.baseUrl, null);
      });
    });
  });
}
